File: CleanJavaRx/app/src/main/java/com/vascome/fogtail/presentation/main/fragment/list/adapter/ListAreaAdapter.java
package com.vascome.fogtail.presentation.main.fragment.list.adapter;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.vascome.fogtail.R;
import com.vascome.fogtail.data.network.AppImageLoader;
import com.vascome.fogtail.presentation.main.domain.model.RecAreaItem;
import com.vascome.fogtail.presentation.main.utils.CollectionAreaItemListener;

import java.util.List;

import static java.util.Collections.emptyList;
import static java.util.Collections.unmodifiableList;

/**
 * Created by vasilypopov on 12/6/17
 * Copyright (c) 2017 MVPJava. All rights reserved.
 */

public class ListAreaAdapter extends RecyclerView.Adapter<ListAreaAdapter.ViewHolder> {


    @NonNull
    private final LayoutInflater layoutInflater;

    @NonNull
    private final AppImageLoader imageLoader;

    private final CollectionAreaItemListener listAreaItemListener;

    @NonNull
    private List<RecAreaItem> items = emptyList();

    public ListAreaAdapter(@NonNull LayoutInflater layoutInflater,
                           @NonNull AppImageLoader imageLoader,
                           CollectionAreaItemListener listAreaItemListener) {
        this.layoutInflater = layoutInflater;
        this.imageLoader = imageLoader;
        this.listAreaItemListener = listAreaItemListener;
    }

    public void setData(@NonNull List<RecAreaItem> items) {
        this.items = unmodifiableList(items); // Prevent possible side-effects.
        notifyDataSetChanged();
    }


    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new ViewHolder(layoutInflater.inflate(R.layout.list_item, parent, false), imageLoader);
    }

    @Override
    public void onBindViewHolder(ViewHolder viewHolder, int position) {
        viewHolder.bind(items.get(position));
        viewHolder.itemView.setOnClickListener(v-> listAreaItemListener.onItemClick(items.get(position)));
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder {

        @NonNull
        private final AppImageLoader imageLoader;

        private final ImageView imageView;

        private final TextView titleTextView;

        private final TextView shortDescriptionTextView;

        ViewHolder(@NonNull View itemView, @NonNull AppImageLoader imageLoader) {
            super(itemView);
            this.imageLoader = imageLoader;
            imageView = itemView.findViewById(R.id.list_item_image_view);
            titleTextView = itemView.findViewById(R.id.list_item_title);
            shortDescriptionTextView = itemView.findViewById(R.id.list_item_description);
        }

        void bind(@NonNull RecAreaItem item) {
            imageLoader.downloadInto(item.imageUrl(), imageView);
            titleTextView.setText(item.name());
            shortDescriptionTextView.setText(item.shortDescription());
        }

    }
}

File: MVVMKotlinArchComponent/app/src/main/java/com/vascome/fogtail/presentation/main/fragment/list/adapter/ListAreaAdapter.kt
package com.vascome.fogtail.presentation.main.fragment.list.adapter

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView

import com.vascome.fogtail.R
import com.vascome.fogtail.data.network.AppImageLoader
import com.vascome.fogtail.presentation.main.dto.RecAreaItem
import com.vascome.fogtail.presentation.main.utils.CollectionAreaItemListener

import java.util.Collections.emptyList
import java.util.Collections.unmodifiableList

/**
 * Created by vasilypopov on 12/6/17
 * Copyright (c) 2017 MVPJava. All rights reserved.
 */

internal class ListAreaAdapter(private val layoutInflater: LayoutInflater,
                      private val imageLoader: AppImageLoader,
                      private val listAreaItemListener: CollectionAreaItemListener) : RecyclerView.Adapter<ListAreaAdapter.ViewHolder>() {

    private var items = emptyList<RecAreaItem>()

    fun setData(items: List<RecAreaItem>) {
        this.items = unmodifiableList(items) // Prevent possible side-effects.
        notifyDataSetChanged()
    }


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(layoutInflater.inflate(R.layout.list_item, parent, false), imageLoader)
    }

    override fun onBindViewHolder(viewHolder: ViewHolder, position: Int) {
        viewHolder.bind(items[position])
        viewHolder.itemView.setOnClickListener { _ -> listAreaItemListener.onItemClick(items[position]) }
    }

    override fun getItemCount(): Int {
        return items.size
    }

    internal inner class ViewHolder(itemView: View, private val imageLoader: AppImageLoader) : RecyclerView.ViewHolder(itemView) {

        private val imageView: ImageView = itemView.findViewById(R.id.list_item_image_view)

        private val titleTextView: TextView = itemView.findViewById(R.id.list_item_title)

        private val shortDescriptionTextView: TextView = itemView.findViewById(R.id.list_item_description)

        fun bind(item: RecAreaItem) {
            item.imageUrl?.let { url -> imageLoader.downloadInto(url,imageView) }
            titleTextView.text = item.name
            shortDescriptionTextView.text = item.shortDescription
        }

    }
}
